### DISCUSSION TITLE
Registering dependent services

### DISCUSSION THREAD

**Discussion by jaredpar** [dotnet/runtime#39863] [General] [answered]
Have a service in my ASP.Net core app that depends on another service and I'd like to expose them via dependency injection. What I ended up settling on is the following:

```cs
services.AddScoped<Service1>(() => 
  var service1 = ...;
  // configure
  return service1;
});

services.AddScoped<Service2>(serviceProvider => 
  var service2 = ...
  service2.Service1 = (Service1)serviceProvider.GetService(typeof(Service1));
  return service2;
});
```

That feels a bit wrong though because it's not declarative and pretty much all of the DI samples for ASP.NET have been declarative. I feel like I should be looking for an overload if `AddScoped` where I can put `Service1` into the parameter list but I'm not seeing one. 

Is this the correct approach for this pattern?

Note: I realize I could make this declarative by defining a third type say `ServiceBag` that has a ctor which takes `Service1` as a parameter and use that to construct `Service2`. That feels equally wrong though because it's just creating a simple container type to make this situation declarative. The type otherwise adds no value.

**Comment by davidfowl** [comment, marked as answer]
Why aren't you using constructors?

**Reply by jaredpar** [reply]
How would constructors help here? Are they automatically considered by the DI system here?

**Reply by jaredpar** [reply]
Ah I see. What tripped me up here is I was thinking of my types in two different ways:

1. Those I specifically designed to be used in DI. Those used the ctor pattern and dependent services connect automatically. 
1. Those I designed outside of DI and, even though they have ctors that can work, I just didn't think about the fact it would automatically hook up here. 

Think part of the reason I made this mental lapse is that my brain is still stuck in MEF land where you tend to opt constructors into DI (item 1 above). Hence I tend to have the mental split of services as stated above. 

So in some ways I'm unstuck here cause my code will just work. Curious though about the question still. Suppose there was a case where ctors didn't work. What is the best pattern for combining separate services together into another one?

**Reply by davidfowl** [reply]
> So in some ways I'm unstuck here cause my code will just work. Curious though about the question still. Suppose there was a case where ctors didn't work. What is the best pattern for combining separate services together into another one?

It's a broken pattern without constructors. The DI system creates a blueprint for constructing types based on constructors (some others do support property injection though but not the built in one). The pattern you show is the only real way to do it (manual calls to serviceProvider.GetService\<T\>) with the built in container.
